content/blog/2020-12-16-compcert/index.md

```diff
@@ -0,0 +1,249 @@
+
+The __tenth pass__ performs constant propagation. This pass may also replace instruction operators in the case that some arguments are statically known. Since the number of transitions may change, correctness needs to take this into account. 
+
+The __eleventh pass__ executes common subexpression elimination, although it's not performed on function calls and it doesn't consider `x + y` and `y + x` equivalent. 
+
+The __twelfth pass__ is dead code elimination. This may make a nonterminating program terminate. For example, suppose a variable is assigned to an integer divided by zero, but that variable is never used. This pass will turn the variable assignment into a `nop` with the justification that the source program was unsound.
+
+The __thirteenth pass__ removes unused global variables. 
+
+The __fourteenth pass__ takes the optimized RTL CFG and transforms it into an [LTL][ltl] CFG, which uses machine registers and stack slots and has basic blocks as nodes instead of individual instructions. Register allocation is done via the [iterated register coalescing algorithm][irca]. The algorithm is implemented in OCaml; if it succeeds, the LTL AST is [verified correct][vregalloc] with respect to the original function.
+
+The __fifteenth pass__ does branch tunneling, which makes uninterrupted branch and conditional branch sequences fall through to the last destination in the sequence. Again, correctness needs to be reevaluated given that the number of transitions may change. 
+
+The __sixteenth pass__ linearizes the LTL CFG by replacing graph edges with jumps (unless a jump would be equivalent to falling through to the following instruction). In order to minimize the number of jumps inserted, an OCaml program picks a block ordering that maximizes fall-throughs. A Coq program then goes over the blocks in that order to generate an instruction sequence. Since program transitions may change in this translation, correctness needs to be adjusted here too.
+
+The __seventeenth pass__ removes unused labels. 
+
+The __eighteenth pass__ modifies the linearized code at call and return points to conform to the target architecture's conventions.
+
+The __nineteenth pass__ outputs an AST corresponding to one of the target assembly languages. 
+
+### Getting to Executable Assembly
+
+An OCaml function prints the target assembly AST with proper register and instruction names. This output is then assembled into an object file with an assembler, and linked with other object files (separately compiled or in the same pass). `gcc` is the most common external assembler and linker used to do this. These three passes are unverified. 
+
+# CompCert Correctness Claims
+
+CompCert promises that it will not mistranslate code it compiles. Specifically, the executable it generates will not perform any action that the source code could not perform. 
```

Comment:
But, as mentioned above, it is not guaranteed to translate every program. So _if_ CompCert produces a compiled program, _then_ it is guaranteed to be correct. Not that this is a problem in practice, I think.

Reply:
I thought that the first sentence drove that home; I've adjusted this paragraph to be more explicit about this.